### DISCUSSION TITLE
Is there any successful implementation of Deep Reinforcement Learning agents in MESA ?

### DISCUSSION THREAD

**Discussion by mrak31** [mesa/mesa#1042] [Q&A] [answered]
Hi, I'm wondering if there is any successful implementation of Deep Reinforcement learning agents in MESA.

There are various Reinforcement learning (RL) libraries / baselines to choose from i.e. (https://blog.dataiku.com/on-choosing-a-deep-reinforcement-learning-library).

Is it worth the effort to implement deep reinforcement learning agents in a MESA based models (simulation environments) ? Eventhough GymAi environments are available - although i noticed the focus of GymAi simulation environment (model) are mostly on single agents.

**Comment by rht** [comment]
I think it is straightforward to implement any Mesa model as a gym environment that can be plugged in with any DRL libraries you use, like RLib. You just need to expose your model `step()` function. No extra setup need to be done from Mesa (the library)'s end.

**Reply by mrak31** [reply]
Hoping there was a specific project someone can point to, but thanks

**Comment by tpike3** [comment, marked as answer]
I can't give you a specific repo but here is a list of repos that use Mesa there maybe one in there https://github.com/projectmesa/mesa/network/dependents 

If you do find please post

**Reply by mrak31** [reply]
Looking into this one:
https://github.com/eescriba/smart-cities-drl

Looks related to what i'm looking for

**Reply by jackiekazil** [reply]
@mrak31 Thank you for highlighting it! Now I am looking into it!

**Comment by rht** [comment]
I just discovered [whynot](https://github.com/mrtzh/whynot).

**Reply by jackiekazil** [reply]
Can you provide more details as to what you are thinking?

**Reply by rht** [reply]
Upon reading further, it supports reinforcement learning only for the existing OpenAI `gym` environments. I thought it had wrapper for Mesa models as an environment for `gym`. Or perhaps @millerjohnp would know better.
